toolkit/offchain/src/reserve/handover.rs

```diff
@@ -36,5 +56,214 @@ pub async fn handover_reserve<
+			validator.bytes.len(),
+		),
+		&Redeemer::new(&RedeemerTag::new_spend(), &0u32.into(), &redeemer_data, &cost),
+	);
+	inputs.add_plutus_script_input(&witness, &input, &amount);
+
+	Ok(inputs)
+}
+
+fn burn_reserve_token(
+	tx_builder: &mut TransactionBuilder,
+	reserve: &ReserveData,
+	costs: ExUnits,
+) -> Result<(), JsError> {
+	let mut mint_builder = tx_builder.get_mint_builder().unwrap_or(MintBuilder::new());
+	let auth_policy = &reserve.scripts.auth_policy;
+	let auth_policy_version_utxo_id = reserve.auth_policy_version_utxo.to_csl_tx_input();
+	let auth_policy_source_ref_input = PlutusScriptSource::new_ref_input(
+		&auth_policy.csl_script_hash(),
+		&auth_policy_version_utxo_id,
+		&auth_policy.language,
+		auth_policy.bytes.len(),
+	);
+	let mint_witness = MintWitness::new_plutus_script(
+		&auth_policy_source_ref_input,
+		&Redeemer::new(
+			&RedeemerTag::new_mint(),
+			&0u32.into(),
+			&PlutusData::new_empty_constr_plutus_data(&0u32.into()),
+			&costs,
+		),
+	);
+	mint_builder.add_asset(&mint_witness, &empty_asset_name(), &Int::new_i32(-1))?;
+	tx_builder.set_mint_builder(&mint_builder);
+	Ok(())
+}
+
+// Creates output with reserve token and updated deposit
+fn illiquid_supply_validator_output(
+	amount: &TokenAmount,
+	scripts: &ReserveScripts,
+	ctx: &TransactionContext,
+) -> Result<TransactionOutput, JsError> {
+	let mut ma = MultiAsset::new();
+	let mut assets = Assets::new();
+	assets.insert(
+		&cardano_serialization_lib::AssetName::new(amount.token.asset_name.0.to_vec())?,
+		&amount.amount.into(),
+	);
+	ma.insert(&amount.token.policy_id.0.into(), &assets);
+
+	let amount_builder = TransactionOutputBuilder::new()
+		.with_address(&scripts.illiquid_circulation_supply_validator.address(ctx.network))
+		.with_plutus_data(&PlutusData::new_empty_constr_plutus_data(&BigNum::zero()))
+		.next()?;
+	let output = amount_builder.with_coin_and_asset(&0u64.into(), &ma).build()?;
+	let ada = MinOutputAdaCalculator::new(
+		&output,
+		&DataCost::new_coins_per_byte(&ctx.protocol_parameters.min_utxo_deposit_coefficient.into()),
+	)
+	.calculate_ada()?;
+
+	amount_builder.with_coin_and_asset(&ada, &ma).build()
+}
+
+struct ScriptsCosts {
+	reserve_auth_policy_spend: ExUnits,
+	reserve_auth_policy_mint: ExUnits,
+	governance_mint: ExUnits,
+}
+
+impl ScriptsCosts {
+	fn zero() -> Self {
+		Self {
+			reserve_auth_policy_spend: zero_ex_units(),
+			reserve_auth_policy_mint: zero_ex_units(),
+			governance_mint: zero_ex_units(),
+		}
+	}
+
+	fn from_ogmios(
+		response: Vec<OgmiosEvaluateTransactionResponse>,
+		reserve_auth_policy_script: &PlutusScript,
+		governance_policy_script: &PlutusScript,
+	) -> Result<Self, anyhow::Error> {
+		let ScriptExUnits { mut mint_ex_units, mut spend_ex_units } =
+			get_validator_budgets(response);
+		let reserve_auth_policy_spend = spend_ex_units
+			.pop()
+			.ok_or_else(|| anyhow!("Evaluate response does not have expected 'spend' cost"))?;
+		let mint_1 = mint_ex_units
+			.pop()
```

Comment:
Ogmios assigns indexes "mints" in the order from mints array from transaction body. CSL sorts mints by script has.

This code makes the assumptions above, so it is simpler compared to handling evaluation responses in other places in the code.

It could be shorter by few lines using some slice deconstruction, but I prefer `pop` api.